### DISCUSSION TITLE
Enforce getIsEditable for all element types

### DISCUSSION THREAD

**Discussion by timkelty** [craftcms/cms#8658] [Ideas]
Despite their generic names, it seems that `ElementHelper::editableSiteIdsForElement` and `craft\base\Element::getIsEditable` are really only concerned with the double-click for HUD edit functionality.

Not knowing this, I ran into trouble when I tried to use it on an element such as `MatrixBlock`, which despite logically being "editable", `MatrixBlock::getIsEditable` is `false`.

To get around this, I had to make a special case for different element types: https://github.com/timkelty/craftcms-site-sync/blob/master/src/models/Syncable.php#L210-L221

What would be ideal is if `ElementHelper::editableSiteIdsForElement` worked on any element type, letting you know which sites the user had permissions to edit the given element.

For this to work with third party plugins, etc. – it would probably make sense to require elements to implement a `getIsEditable`.

`MatrixBlock` might then have something like this:

```php
    public function getIsEditable(): bool
    {
        return $this->owner->getIsEditable();
    }
```

**Comment by brandonkelly** [comment]
Not fully resolved yet, but in Craft 3.4, element index pages will stop adding hyperlinks to elements if `getIsEditable()` returns `false`.
